packages/app-desktop/gui/NoteListControls/NoteListControls.tsx

```diff
@@ -208,7 +208,10 @@ function NoteListControls(props: Props) {
 
 		return (
 			<TopRow className="new-note-todo-buttons">
-				<StyledButton ref={props.newNoteButtonRef}
+				<StyledButton
+					ref={(el: Element) => {
+						props.setNewNoteButtonRef(el);
+					}}
```

Comment:
Sorry I don't know why it works but this is not the right pattern to set references, so I'd rather not go with this change.

You also changed the logic to update the width or not, so maybe you fixed it this way and this change to the ref doesn't do anything

Reply:
Hi, thanks for your feedback. I read from here: [render when ref change](https://stackoverflow.com/questions/59600572/how-to-rerender-when-refs-change ). I also know it recently, it's a pattern in React [callback ref](https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs). 

Also, i didnt change the change width logic, usually newNoteButton.current is the element, but im using state, so newNoteButton is the element, so it stills same element just different way to access

Reply:
My idea is to change from ref to state, so it can be re-render, because using ref as a dependency in useEffect doesnt trigger it or the useRef hang on old value (I tested this, and this is how React designed useRef too). 

To sum up: I want to use State instead of Ref, maybe my implementation is weird, I'd appreciate if you guide me some part of it, thank you

Reply:
I can prove that changing width logic doesn't fix the problem, but it's the state vs ref  that fixes.

Reply:
If it works now I think it's only by chance, as a side effect of forcing a rendering when the state is updated.

It would be better to fully understand why it doesn't render.  Usually it means that the effect is not taking into account a particular state variable. For example before it wasn't taking into account the width of the note list so that's why I've added this `widthHasChanged` check.

So I think the ref is fine but we need to find what part of the application state is not being taken into account.

Reply:
# useRef version
This is total old version, I just log newNoteRef. You see, it hang on to old value.

https://github.com/laurent22/joplin/assets/42113313/e6e103fd-efd8-48ed-bb3b-3932156de8d1

And this is from [React useRef() docs](https://legacy.reactjs.org/docs/hooks-reference.html#useref)
![image](https://github.com/laurent22/joplin/assets/42113313/7ab2c73f-e190-4863-aa20-c1c3699657ab)

My understanding:  when switch to another tab, showNewNoteButton triggers useEffect, but newNoteRef stills hang on previous value. When newNoteRef change to new value, it doesn't trigger useEffect for re-rendering (because of useRef behavior from React docs above). 

# useState version
This is my useState version, every click to tab, run the useEffect twice.

https://github.com/laurent22/joplin/assets/42113313/52c849bb-750f-4661-9972-8b39d8d83d9b

My understanding: when switch to another tab, showNewNoteButton triggers useEffect once, but newNoteRef stills hang on previous value (1st print is oldState). When newNoteRef change to new value, it triggers useEffect again  (2nd print is newState). This keep the newNoteRef in sync with the updating breakpoints.

Also, the whole thing I do just switching tabs, no change width is involved.